### ISSUE TITLE
ALPN filter incorrectly applies to non-Istio TLS traffic

### ISSUE THREAD

**Issue by infa-ddeore** [istio/istio#40680]
### Bug Description

setup:
- ssl terminating on istio gateway
- backend app listens on https with istio sidecar injected
- destination rule has tls mode as SIMPLE

curl accessing gateway's https port:

1.  with 1.13.x
- works
2. with 1.14.x
- fails with `upstream connect error or disconnect/reset before headers. reset reason: connection termination` error
- works - when mTLS is disabled in peer authentication
- works - when istio sidecar isnt injected to the app

### Version

```prose
$ istioctl version
client version: 1.14.1
istiod version: 1.13.5
istiod version: 1.14.1
data plane version: 1.14.1 (1 proxies)

$ k version --short
Client Version: v1.23.5
Server Version: v1.21.11
WARNING: version difference between client (1.23) and server (1.21) exceeds the supported minor version skew of +/-1
```

### Additional Information

a very basic [golang server](https://github.com/deepakdeore2004/istio-bug-1-14/blob/main/server.go) is used to reproduce the problem

how to reproduce:

- install istio 1.13 and 1.14 as canary revision
- install `istioctl` (1.13.x or 1.14.x) on local and make it available in PATH to deploy gateway
- clone git@github.com:deepakdeore2004/istio-bug-1-14.git and change directory to k8
- ./run.sh, provide istio revision to deploy gw and app with istio sidecar
- curl -k https://gateway-lb
- delete the namespace and repeat same test with another revision

in 1.14, disabling mTLS with peer authentication OR removing istio sidecar from the app works fine

alpn has some difference with the version

1.13:

<img width="279" alt="1 13" src="https://user-images.githubusercontent.com/39782583/186964575-38707a5b-7fc4-493c-86f5-b95bad877f8b.png">

1.14:
<img width="327" alt="1 14" src="https://user-images.githubusercontent.com/39782583/186964580-0913a53f-a35e-4347-9fd2-ec3cde7b3ae2.png">

tcpdump is installed on app pod so kubectl sniff can be used like this:
```
k sniff -n test-1-13-5-20220827 test-http-server-8080-and-8443-746f8679f5-d6sgl
```

**Comment by hzxuzhonghu** [comment]
Looks this is expected, I cannot recalll what cause this diff between 1.13 and 1.14. @howardjohn @ramaraochavali do you know what has changed

**Comment by howardjohn** [comment]
I think I know what happens, its not what I expected originally.. and I am also pretty sure its broken in 1.13 as well.

Basically Istio inserts the istio.alpn filter for HTTP traffic into the HCM. The purpose of this is to dynamically set the ALPN from things like h2 -> istio-h2, http/1.1->istio-http/1.1, etc. On the server side, we detect these `istio-x` variants to identify if its mTLS. We need the `istio-http` ones (vs just `istio`) to distinguish HTTP vs TCP for auto sniffed protocols - in theory we could still sniff after we terminate TLS, but the Envoy config model doesn't really support that, so we need the client to tell us what protocol it is.

The problem arises when we _aren't_ sending mTLS, we still apply the ALPN filter. This is unavoidable in the current implementation, since its applied at the HCM level (shared), but we need to enable/disable on a per-cluster basis.

The 1.14 regression was just making this broken for gateway -> sidecar as well; it was always broken for sidecars.

**Comment by howardjohn** [comment]
Possible options:

1. Make the ALPN filter more expressive to allow selectively disabling (or apply to cluster directly somehow)
2. Stop with the `istio-httpX` stuff and just pick one of them (say `istio-http/1.1`). We never actually distinguish between the 3 version variants - always match on all of them together. With this, we can just drop the filter entirely and set the ALPN directly in the cluster, I believe (not validated whether this will work). One downside is we would be "lying" and always declaring `http/1.1` (or always `h2`), when we are not actually using those protocols
3. something else?

**Comment by howardjohn** [comment]
@briansonnenberg is going to work on this but I cannot assign them since they are not in Istio org yet

**Comment by kyessenov** [comment]
The real solution is using HBONE for Istio TLS. Custom ALPN is a true abuse and unlikely to receive support from upstream Envoy. It needs to be deprecated.

A workaround for per-cluster selection is to use `StreamInfo::clusterInfo()` in the filter to adjust its behavior. That will break with filters that change routes, but might help temporarily.

**Comment by howardjohn** [comment]
HBONE will only solve this once non-HBONE mTLS is _entirely_ removed, IIUC. Which is likely to be years from now due to our backwards compatibility requirements

**Comment by kyessenov** [comment]
@howardjohn we have fairly short support window and Istio is the only data plane that uses these ALPNs. I think we need to start accepting HBONE traffic ASAP, and then take deprecation cycle to stop sending on HBONE traffic, but that doesn't seem to be anything special.

**Comment by kyessenov** [comment]
@costinm for thoughts. I really don't dig a deeper hole for the ALPNs hacks if we can avoid it.

**Comment by howardjohn** [comment]
We have almost never broken data plane version skew compatibility historically. Even today we still upgrade users directly from Istio 1.4 to Istio 1.15 and have those proxies communicate.

**Comment by kyessenov** [comment]
If we truly want compatibility then unfortunately we can't fix bug or any other data plane either :) Whoever runs 1.4 proxy is already taking massive amount of risks with public CVEs.

**Comment by howardjohn** [comment]
1.4 is just an example. Even if we have the strictest interpretation of backwards compat and only support skew between supported versions, we will still have mTLS around for at least a year. I suspect we will end up needing a longer period of time though, since HBONE will not be stable on initial merge. So we have at minimum ~6 months to HBONE as a requirement + 1 year to deprecate mTLS most likely.

Bug fixing is largely not an issue on version skew. The shared protocol between two proxies is fairly small - TLS with a shared root, and our ALPN hack. Maybe MX if we also want to make sure telemetry works as well. The version skew support is not around ensuring the behavior of an individual proxy is identical between versions (which is a far higher bar), just that 2 proxies can communicate

**Comment by kyessenov** [comment]
I'm not suggesting to drop to the support for old ALPN hack before the regular deprecation cycle. But this bug could be avoided by using a tunnel, so if we support that as an option, it seems like a no-brainer to invest into that rather than fixing the old path. We'd have to run dual protocols in any case, and there's no problem listening for both on the sidecars.

**Comment by howardjohn** [comment]
I don't see how it can be avoided by introducing HBONE without dropping old mTLS. The issue is the config is at the HCM level, and the HCM is shared by ~everything, so the filter must be present which will lead to overwriting the ALPN.

I think it will actually impact HBONE as well - we will be using our ALPN hacks _in_ HBONE because we cannot selectively enable/disable it at a cluster level.

**Comment by kyessenov** [comment]
The effort should be on supporting HBONE. If it takes changes to istio_alpn filter to opt-out, then that should be the work. There are many things that need to happen for HBONE besides that.

**Comment by howardjohn** [comment]
Ok, well I think it turns out the changes need to support HBONE _and_ to fix this bug are the same, so I think that makes everyone happy.

**Comment by infa-ddeore** [comment]
@howardjohn this needs fix or can be closed?

**Comment by howardjohn** [comment]
Needs a fix

**Comment by keithmattix** [comment]
I've put together a Design Doc for fixing this: https://docs.google.com/document/d/1T8NvwK2hnRjU-LWWtZ_Wk6kIVkeZqYEcP_LQ3dEkO-8/edit?usp=sharing. Hopefully we can discuss at the next WG meeting

**Comment by jewertow** [comment]
Despite recent changes in istio-proxy and istiod, I am still able to reproduce this issue. I also noticed that this issue can be reproduced in Istio 1.13 when STRICT PeerAuthention is enabled. To fix it, it's necessary to disable mTLS on app's TLS port. What's more, disabling mTLS with and without global STRICT PeerAuthentication solves the issue in Istio 1.18.
```
version | resources                                | TLS handshake
------- + ---------------------------------------- + --------------
1.13    | Gateway (SIMPLE)                         | succeeds
        | DestinationRule (SIMPLE)                 |
------- + ---------------------------------------- + --------------
1.18    | Gateway (SIMPLE)                         | fails
        | DestinationRule (SIMPLE)                 |
------- + ---------------------------------------- + --------------
1.13    | Gateway (SIMPLE)                         | fails
1.18    | DestinationRule (SIMPLE)                 |
        | PeerAuthentication (STRICT/istio-system) |
------- + ---------------------------------------- + --------------
1.13    | Gateway (SIMPLE)                         | succeeds
1.18    | DestinationRule (SIMPLE)                 |
        | PeerAuthentication (STRICT/istio-system) |
        | PeerAuthentication (DISABLE/app selector)|
------- + ---------------------------------------- + --------------
```
Taking the above into account, I wonder if we really need those recent changes in istio.alpn filter, because it seems to me that disabling PeerAuthentication solves the problem and has the same limitations - no peer metadata exchange. And if disabling alpn filter does not fix TLS handshake when STRICT PeerAuthentication is enabled, does this change still make sense?
cc: @ksubrmnn @keithmattix @howardjohn

**Comment by keithmattix** [comment]
Hmm this tells me that the Envoy PeerAuthentication implementation makes an assumption that mTLS == Istio mTLS which IMO is a faulty assumption. What I think should happen is that PeerAuthentication mtls == STRICT should make the target services have `require_client_certificate` set to true.

**Comment by howardjohn** [comment]
STRICT mTLS means it must be Istio mutual TLS. In your case 3, its not Istio mutual TLS -- its 'application level' TLS and is correctly rejected.

**Comment by keithmattix** [comment]
If STRICT mTLS is only Istio mutual TLS, then we need to update our public-facing docs and type comments because I don't see that anywhere: https://istio.io/latest/docs/reference/config/security/peer_authentication/

**Comment by hzxuzhonghu** [comment]
STRICT does not equal istio mtls

**Comment by howardjohn** [comment]
I am a bit confused ... maybe terminology issue. Let me be more explicit.

First the docs. They say "Connection is an mTLS tunnel" which is a bit ambiguous but does clearly rule out "SIMPLE" TLS.

Next is implementation. Strict will check a incoming connecting is TLS and has a valid SPIFFE mesh certificate. The only variance from "Istio mutual" TLS is that I don't think it enforced the ALPN.

**Comment by keithmattix** [comment]
> They say "Connection is an mTLS tunnel" which is a bit ambiguous but does clearly rule out "SIMPLE" TLS.

Agreed - I'm realizing now I misunderstood the implications of @jewertow's experiment. SIMPLE TLS should, in fact, be blocked by an mTLS STRICT `PeerAuthentication`. However, I do think that my message is correct in form if not in function. More on that below:

> Strict will check a incoming connecting is TLS and has a valid SPIFFE **mesh certificate**

This is the disconnect. The API for PA states is all about mTLS specifically, meaning a client cert must be presented. This diverges from @jewertow's original experiments, but If DestinationRule MUTUAL is set (vs. ISTIO_MUTUAL), then a STRICT PA should not block the traffic.

**Comment by howardjohn** [comment]
I feel quite confident that the correct behavior is to verify as I specified. It is both logically correct (if not that verification... what are you verifying?) and the existing behavior since Istio inception.

The doc is misleading because we mostly interchangablely use mTLS to mean "Istio mTLS" incorrectly, and generally intermix application layer and mesh transport layer everywhere. But it's a doc issues not behavioral.

**Comment by howardjohn** [comment]
I mean if they set MUTUAL and then present an Istio mesh certificate sure.. but then that is really "Istio mTLS" done manually; doesn't change the actual behavior on the server side

**Comment by hzxuzhonghu** [comment]
I think we need to update the `inboundStrictFilterChainMatchOptions`,. As with a DR specify simple or MTLS mode, we will not override alpn now. So the inbound alpn is ["http/1.1", "h2c"], so we could not match ["istio-http/1.0", "istio-http/1.1", "istio-h2"]

**Comment by keithmattix** [comment]
@howardjohn fair enough re: the docs. My position is based on my reading of the docs only, so I'll defer to you on the history and intent. We should definitely correct the docs though.

**Comment by howardjohn** [comment]
> I think we need to update the `inboundStrictFilterChainMatchOptions`,. As with a DR specify simple or MTLS mode, we will not override alpn now. So the inbound alpn is ["http/1.1", "h2c"], so we could not match ["istio-http/1.0", "istio-http/1.1", "istio-h2"]

https://github.com/dubbo-go-pixiu/dubbo-go-pixiu/blob/53014058bac52d74fd98ba94a0a8b0e2c2b92c3e/pilot/pkg/networking/core/v1alpha3/filterchain_options.go#L127 it already is the istio- ones

**Comment by hzxuzhonghu** [comment]
But we donot do alpn overwrite now, so it cannot match the filterchain.

**Comment by howardjohn** [comment]
We do the ALPN overwrite - the change was only to not do it for non-ISTIO_MUTUAL. It doesn't make sense to set the ALPN overwrite for SIMPLE since it's not mTLS?

**Comment by jewertow** [comment]
When I enabled debug logs in ingress gateway, I noticed `override with 3 ALPNs` even though the cluster had `istio.alpn_override: false`, so I think current implementation of skipping ALPN override is not correct. I submitted a fix: https://github.com/istio/proxy/pull/4865.

**Comment by kyessenov** [comment]
Is there a summary for why Istio needs to have a fork of mTLS for its mesh transport? It's really not ideal and seems to be a historic artifact that we should be working to eliminate, instead of promoting further. HBONE would get us there quickly, but it's been multiple years waiting for it, so we should probably not expect it to happen soon, and make incremental changes. From what I remember:

* istio-metadataexchange ALPN is metadata exchange ALPN. We can get rid of it with WDS in the upcoming version.
* istio-h2, etc: is this about sniffing on the client? Envoy can sniff on the server now, so not sure why we can't sniff twice.

Anything else?

**Comment by keithmattix** [comment]
Can you explain what you mean by a "fork" of mTLS?

**Comment by kyessenov** [comment]
By that I mean that we need to distinguish between any mTLS and "istio mTLS" - that's by definition a new protocol based of mTLS. A client speaking regular mTLS (e.g. gRPC) cannot access the sidecar directly. The fact that we define new non-standard ALPN tokens  (not present in https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids) means Istio requires a sidecar to be in the mesh.

**Comment by howardjohn** [comment]
Technically we don't check the ALPN in STRICT mode so you can use it in that case.

> Anything else?

A major purpose is distinguishing "application TLS" from "mesh TLS".

**Comment by kyessenov** [comment]
> A major purpose is distinguishing "application TLS" from "mesh TLS".

So the only reason is because we cannot sniff istio-enabled vs application TLS in "permissive mode" on the server? why isn't auto-mTLS solving this?

**Comment by howardjohn** [comment]
> > A major purpose is distinguishing "application TLS" from "mesh TLS".
> 
> So the only reason is because we cannot sniff istio-enabled vs application TLS in "permissive mode" on the server? why isn't auto-mTLS solving this?

That only works if the clients are in the mesh and have auto-mTLS activated (it can be "not activated" by turning it off globally, a DISABLE DR, or not going through an in-scope Service (direct to pod, exportTo, sidecar scope, etc)).

**Comment by infa-ddeore** [comment]
> When I enabled debug logs in ingress gateway, I noticed `override with 3 ALPNs` even though the cluster had `istio.alpn_override: false`, so I think current implementation of skipping ALPN override is not correct. I submitted a fix: [istio/proxy#4865](https://github.com/istio/proxy/pull/4865).

which istio version has this fix?

**Comment by keithmattix** [comment]
I think this landed in 1.19
